pkg/agent/attestor/node/node.go

```diff
@@ -81,37 +82,54 @@ func (a *attestor) Attest(ctx context.Context) (res *AttestationResult, err erro
 	return &AttestationResult{Bundle: bundle, SVID: svid, Key: key}, nil
 }
 
+// Load the current SVID and key. The returned SVID is nil to indicate a new SVID should be created.
 func (a *attestor) loadSVID(ctx context.Context) ([]*x509.Certificate, *ecdsa.PrivateKey, error) {
 	km := a.c.Catalog.GetKeyManager()
-	fResp, err := km.FetchPrivateKey(ctx, &keymanager.FetchPrivateKeyRequest{})
+	fetchRes, err := km.FetchPrivateKey(ctx, &keymanager.FetchPrivateKeyRequest{})
 	if err != nil {
 		return nil, nil, fmt.Errorf("load private key: %v", err)
 	}
-
 	svid := a.readSVIDFromDisk()
-	if len(fResp.PrivateKey) > 0 && svid == nil {
-		a.c.Log.Warn("Private key recovered, but no SVID found")
-	}
 
-	var keyData []byte
-	if len(fResp.PrivateKey) > 0 && svid != nil {
-		keyData = fResp.PrivateKey
-	} else {
-		gResp, err := km.GenerateKeyPair(ctx, &keymanager.GenerateKeyPairRequest{})
+	privateKeyExists := len(fetchRes.PrivateKey) > 0
+	svidExists := svid != nil
+	svidIsExpired := isSVIDExpired(svid, time.Now)
+
+	switch {
+	case privateKeyExists && svidExists && !svidIsExpired:
+		key, err := x509.ParseECPrivateKey(fetchRes.PrivateKey)
 		if err != nil {
-			return nil, nil, fmt.Errorf("generate key pair: %s", err)
+			return nil, nil, fmt.Errorf("parse key from keymanager: %v", key)
 		}
-
-		svid = nil
-		keyData = gResp.PrivateKey
+		return svid, key, nil
+	case privateKeyExists && svidExists && svidIsExpired:
+		a.c.Log.Warn("Private key recovered, but SVID is expired. Generating new keypair.")
+	case privateKeyExists && !svidExists:
+		a.c.Log.Warn("Private key recovered, but no SVID found. Generating new keypair.")
+	case !privateKeyExists && svidExists:
+		a.c.Log.Warn("SVID recovered, but no private key found. Generating new keypair.")
+	default:
+		// Neither private key nor SVID were found.
+	}
+
+	generateRes, err := km.GenerateKeyPair(ctx, &keymanager.GenerateKeyPairRequest{})
+	if err != nil {
+		return nil, nil, fmt.Errorf("generate key pair: %s", err)
 	}
-
-	key, err := x509.ParseECPrivateKey(keyData)
+	key, err := x509.ParseECPrivateKey(generateRes.PrivateKey)
 	if err != nil {
 		return nil, nil, fmt.Errorf("parse key from keymanager: %v", key)
 	}
+	return nil, key, nil
+}
 
-	return svid, key, nil
+func isSVIDExpired(svid []*x509.Certificate, timeNow func() time.Time) bool {
+	if len(svid) == 0 {
+		return false
+	}
+	clockSkew := time.Second
```

Comment:
What scenario are you trying to mitigate with adjusting for 1 second of skew?

Reply:
This is a fairly tiny edge case, but I wanted to prevent a case like the following:

1. Current certificate expires at time t=10
2. Agent thinks its t=9, so it uses the current certificate
3. Server time is t=10, so it considers this certificate invalid
4. Agent fails to boot

Even a relatively small clock skew (100ms) could result in this, especially when you account for network latency.

Reply:
What if the server is 1.5 seconds ahead? 2 seconds? 2.5 seconds? :)

There is always a chance the server clock is ahead of the client. I think a more robust handling would be to detect TLS handshake failures due to the server rejecting the certificate and properly fall back to re-attestation.

That being said, it would be a much more involved change, so I'm ok with this as-is, since I believe it is a great step in the right direction and certainly improves the UX in the short term for this problem. Thanks for the explanation!